lib/POMDPTools/src/Policies/dict.jl

```diff
@@ -0,0 +1,60 @@
+### Dict Policy ###
+
+
+"""
+     DictPolicy{P<:Union{POMDP,MDP}, T<:AbstractDict{Tuple,Float64}}
+A generic MDP policy that consists of a value dict.
+
+# Fields 
+- `mdp::P` the MDP problem
+- `value_table::T` the value dict, key is (s,a) Tuple.
+"""
+
+mutable struct DictPolicy{P<:Union{POMDP,MDP}, T<:AbstractDict{Tuple,Float64}} <: Policy
+    mdp::P
+    value_dict::T
+end
+
+function DictPolicy(mdp::Union{MDP,POMDP})
+    return DictPolicy(mdp, Dict{Tuple,Float64}())
+end
+
+# return the action with the max value
+function action(p::DictPolicy, s)
+    available_actions = actions(p.mdp,s)
+    max_action = nothing
+    max_action_value = 0
+    for a in available_actions
+        if haskey(p.value_dict,(s,a))
+            action_value = p.value_dict[(s,a)]
+            if action_value > max_action_value
+                max_action = a
+                max_action_value = action_value
+            end
+        else
+            p.value_dict[(s,a)] = 0
+        end
+    end
+    if max_action === nothing
+        max_action = available_actions[1]
```

Comment:
```suggestion
        max_action = first(available_actions)
```
`first` is preferable to `[1]` because it will work for containers that are not indexable (e.g.)

Reply:
More importantly, I think it would be better to use a default policy here, i.e. store `default` in the struct and call `action(p.default, s)` here. I think it would be even better to have a user-defined default value (implicitly you have defined it with `max_action_value=0` here. Then, if none of the values in the `Dict` are higher, it will use the default policy.

Reply:
I agree with you. So I will add a new field `default_policy` in struct `ValueDictPolicy`.
And I think it would be better if we have two constructors for `ValueDictPolicy`.
`ValueDictPolicy(mdp::MDP)` : Random return one of action from `actions(p.mdp,s)`, if max value action is not exist.
`ValueDictPolicy(mdp::MDP, default_policy::Policy)` : Return action by `default_policy`, if max value action is not exist.

Reply:
Shall we let user define default value?
 I set `max_action_value=0` because I want to keep the same behavior with `ValuePolicy`, the default value of `ValuePolicy` is 0, before we encounter a specific state-action pair in the environment. [Souce Code](https://github.com/JuliaPOMDP/POMDPs.jl/blob/8e7e7f147a99a96799b32c30f431747866c1bc11/lib/POMDPTools/src/Policies/vector.jl#LL58C48-L58C48)

Reply:
> More importantly, I think it would be better to use a default policy here, i.e. store `default` in the struct and call `action(p.default, s)` here.

Do you mean to expose how to select action to the user?
Or just a policy seal in the struct and only execute in `action()` function?

Reply:
> I set max_action_value=0 because I want to keep the same behavior with ValuePolicy, the default value of ValuePolicy is 0, before we encounter a specific state-action pair in the environment. Souce Code

Yes, but in `ValuePolicy` the user can change that if they want to. I think you should have
```julia
struct DictPolicy{M<:MDP, T<:AbstractDict{Tuple,Float64}, P<:Policy} <: Policy
    mdp::M
    value_dict::T
    default_value::Float64
    default_policy::P
end
```
with the constructor setting `default_value=-Inf` unless the user overrides it. And then
```julia
max_action_value = p.default_value
```
initially, and
```julia
if isnothing(max_action)
    return action(p.default_policy)
```
near the end of the function. Does that make sense?

Reply:
I think `default_value` is very helpful, users could control the "initial value" of the "value table" (it's dict here) 

But `default_policy`, it will be available on the situation "there is no max value action, so let default_policy help me decide what action to return" only. In most instances, it's the first time we encounter a state, so all the value is the same.

But shall we let users decide  what action returned, when all the action values are the same? 
I'm not sure, maybe just returning random action is better?

Could you please give me an example about "in this situation we must let users define policy select action"?
Any solver algorithm need control policy here?

Thank you very much, if you have any ideas please let me know, I will finish it.